### ISSUE TITLE
docdiff: highlights unchanged tables, not actual changes

### ISSUE THREAD

**Issue by stevepiercy** [readthedocs/addons#348]
Example PR preview: https://plone6--1679.org.readthedocs.build/install/containers/images/backend.html#arbitrary-user-and-persistent-data

When I use `d` to toggle between docdiff views, the changes from the PR do not appear as a diff, but instead all tables in the page display diff highlighting even though there were no changes to them.

**Comment by humitos** [comment]
Thanks for reporting this issue. I've seen this problem when using docdiff on pages with tables, but I think I never created the issue and then forgot about this 😅 

Since the content for the original version could change over time, I created a small GIF to visually document the issue:

![Peek 2024-07-18 11-47](https://github.com/user-attachments/assets/b85f3b23-6cc5-4382-acb2-7d1ce053a639)

I also downloaded both of the versions for that file, and it's clear there is no changes in the tables at all:

```diff
$ diff -u original.html pr.html 
--- original.html	2024-07-18 03:22:43.000000000 +0200
+++ pr.html	2024-07-17 12:42:39.000000000 +0200
@@ -80,7 +80,7 @@
 </script>
 <!-- End Matomo Tag Manager -->
 
-  <script async type="text/javascript" src="/_/static/javascript/readthedocs-addons.js"></script><meta name="readthedocs-project-slug" content="plone6" /><meta name="readthedocs-version-slug" content="6.0" /><meta name="readthedocs-resolver-filename" content="/install/containers/images/backend.html" /></head>
+  <script async type="text/javascript" src="/_/static/javascript/readthedocs-addons.js"></script><meta name="readthedocs-project-slug" content="plone6" /><meta name="readthedocs-version-slug" content="1679" /><meta name="readthedocs-resolver-filename" content="/install/containers/images/backend.html" /></head>
   <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="100">
 <!-- Checkboxes to toggle the left sidebar -->
 <input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
@@ -519,7 +519,7 @@
      </li>
      <li class="toctree-l3">
       <a class="reference internal" href="../../../volto/configuration/volto-config-js.html">
-       Programmatically define the active add-ons and theme
+       Programatically define the active add-ons and theme
       </a>
      </li>
      <li class="toctree-l3 has-children">
```

I suppose the underlying JavaScript library we use is performing in a weird way here.

**Comment by humitos** [comment]
It seems this issue may be related to https://github.com/Teamwork/visual-dom-diff/issues/7

**Comment by agjohnson** [comment]
I've noticed this happening on sites using sphinx-rtd-theme, and I believe it's because the theme dynamically adds a class to tables after loading -- this is enough for the DOM node to be considered different. This is my theory though, I haven't tested this.

**Comment by humitos** [comment]
I commented on the upstream issue with an example and it seems it's not the case you are describing. There is something else going on here

**Comment by agjohnson** [comment]
Also, note: though this library worked as an initial proof of concept, it was clear from the start that we might have to take on some maintenance of the library to make it work like we need -- it hasn't seen any commits in a years now. Unfortunately, it's Typescript, which will be an added hurdle to this.

This library and a couple of the small bugs like this are one of the reasons I've been hesitant to promote docdiff just yet.

**Comment by humitos** [comment]
Part of this issue was solved by https://github.com/Teamwork/visual-dom-diff/pull/46. Now, tables that are exactly the same won't be considered as changed. However, there are other cases that are not fixed yet: `colspan`, etc. We just need to update the underlying library to 0.7.3.

**Comment by zmbc** [comment]
> I've noticed this happening on sites using sphinx-rtd-theme, and I believe it's because the theme dynamically adds a class to tables after loading -- this is enough for the DOM node to be considered different. This is my theory though, I haven't tested this.

It's not just a class -- the sphinx-rtd-theme actually [wraps all tables in a div element](https://github.com/readthedocs/sphinx_rtd_theme/blob/795de79c8b311592f5863a25307d85924bf52164/src/theme.js#L102-L104), which means they will always appear different between the previous version (on which Javascript is not run) and the current version (on which Javascript is run), due to #528.

> Part of this issue was solved by https://github.com/Teamwork/visual-dom-diff/pull/46. Now, tables that are exactly the same won't be considered as changed.

That PR means that tables that are changed *only by adding newlines* won't be considered as changed. Which is good, but not part of this issue (completely unchanged tables, in which there was no diff in the source, wouldn't randomly add newlines). It was a separate bug that you (@humitos) happened to run into while looking at https://github.com/Teamwork/visual-dom-diff/issues/7#issuecomment-2237049921.

So to summarize the status here, I expect there are currently two reasons why unchanged tables are showing up as changed:
* Certain themes modify the DOM for tables dynamically, which triggers #528 
* Tables with `colspan` and `rowspan` are not handled by the underlying library, see https://github.com/Teamwork/visual-dom-diff/issues/7
